// customisable-contact-form/dist/index.js
function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var reactHookForm = require('react-hook-form');
var styled = require('styled-components');
var styled__default = _interopDefault(styled);

function _taggedTemplateLiteralLoose(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }

  strings.raw = raw;
  return strings;
}

function _templateObject6() {
  var data = _taggedTemplateLiteralLoose(["\n  width: ", ";\n  margin: auto;\n  margin-top: 1em;\n  border: ", " solid\n    ", ";\n  color: ", ";\n  background-color: ", ";\n  padding: ", ";\n  border-radius: ", ";\n  display: flex;\n  flex-direction: column;\n  justify-content: center;\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose(["\n  display: block;\n  box-sizing: border-box;\n  width: 100%;\n  border-radius: ", ";\n  border: ", ";\n  padding: 10px 15px;\n  margin-bottom: 10px;\n  background: grey;\n  font-family: ", ";\n  text-align: center;\n  background-color: ", ";\n  color: ", ";\n  text-transform: uppercase;\n  border: none;\n  margin-top: 30px;\n  padding: ", ";\n  font-size: 20px;\n  font-weight: ", ";\n  display: flex;\n  justify-content: center;\n  letter-spacing: ", ";\n  margin-right: ", ";\n  &:hover {\n    background: ", ";\n  }\n  &:active {\n    transition: 0.3s all;\n    transform: translateY(5px);\n    border: 1px solid transparent;\n    opacity: 0.8;\n  }\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n  line-height: 2;\n  text-align: left;\n  display: block;\n  margin-bottom: 13px;\n  margin-top: 20px;\n  color: ", ";\n  font-size: 14px;\n  font-weight: 200;\n  font-family: ", ";\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n  display: block;\n  box-sizing: border-box;\n  width: 100%;\n  border-radius: ", ";\n  border: ", " solid\n    ", ";\n  padding: 10x 15px;\n  margin-bottom: 10px;\n  font-size: 14px;\n  background-color: ", ";\n  height: ", ";\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n  display: block;\n  box-sizing: border-box;\n  width: 100%;\n  border-radius: ", ";\n  border: ", " solid\n    ", ";\n  padding: 10px 15px;\n  margin-bottom: 10px;\n  font-size: 14px;\n  height: 40px;\n  background-color: ", ";\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  font-weight: ", ";\n  color: ", ";\n  text-align: center;\n  padding-bottom: 10px;\n  border-bottom: ", " solid\n    ", ";\n  font-family: ", ";\n  font-size: ", ";\n  background-color: ", ";\n  border-radius: ", ";\n  display: flex;\n  justify-content: center;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var StyledHeading = styled__default.h1(_templateObject(), function (props) {
  return props.theme.fontWeight;
}, function (props) {
  return props.theme.headingColor;
}, function (props) {
  return props.theme.headingBorderBottom;
}, function (props) {
  return props.theme.headingBorderColor;
}, function (props) {
  return props.theme.primaryFont;
}, function (props) {
  return props.theme.headingSize;
}, function (props) {
  return props.theme.headingBackgroundColor;
}, function (props) {
  return props.theme.headingBorderRadius;
});
var StyledInput = styled__default.input(_templateObject2(), function (props) {
  return props.theme.inputBorderRadius;
}, function (props) {
  return props.theme.inputBorderWeight;
}, function (props) {
  return props.theme.inputBorderColor;
}, function (props) {
  return props.theme.inputBackgroundColor;
});
var StyledMessageInput = styled__default.input(_templateObject3(), function (props) {
  return props.theme.inputBorderRadius;
}, function (props) {
  return props.theme.inputBorderWeight;
}, function (props) {
  return props.theme.inputBorderColor;
}, function (props) {
  return props.theme.inputBackgroundColor;
}, function (props) {
  return props.theme.messageInputHeight;
});
var StyledLabel = styled__default.label(_templateObject4(), function (props) {
  return props.theme.primaryColor;
}, function (props) {
  return props.theme.primaryFont;
});
var StyledButton = styled__default.button(_templateObject5(), function (props) {
  return props.theme.inputBorderRadius;
}, function (props) {
  return props.theme.inputBorder;
}, function (props) {
  return props.theme.primaryFont;
}, function (props) {
  return props.theme.buttonBackgroundColor;
}, function (props) {
  return props.theme.buttonFontColor;
}, function (props) {
  return props.theme.buttonPadding;
}, function (props) {
  return props.theme.fontWeight;
}, function (props) {
  return props.theme.buttonLetterSpacing;
}, function (props) {
  return '-' + props.theme.buttonLetterSpacing;
}, function (props) {
  return props.theme.buttonHoverColor;
});
var StyledForm = styled__default.form(_templateObject6(), function (props) {
  return props.theme.width;
}, function (props) {
  return props.theme.formBorderWeight;
}, function (props) {
  return props.theme.formBorderColor;
}, function (props) {
  return props.theme.primaryColor;
}, function (props) {
  return props.theme.backgroundColor;
}, function (props) {
  return props.theme.formPadding;
}, function (props) {
  return props.theme.formBorderRadius;
});

var defaultTheme = {
  primaryColor: 'black',
  primaryFont: 'Courier New',
  buttonLetterSpacing: '20px',
  buttonPadding: '20px',
  buttonBackgroundColor: 'black',
  buttonFontColor: 'white',
  headingBackgroundColor: '#fbf8f5',
  headingBorderRadius: '',
  headingSize: '30px',
  headingColor: 'black',
  inputBackgroundColor: 'white',
  inputBorderRadius: '15px',
  inputBorderWeight: '2px',
  inputBorderColor: 'grey',
  formPadding: '20px',
  formBorderRadius: '35px',
  formBorderWeight: '1px',
  formBorderColor: 'grey',
  backgroundColor: '#fbf8f5',
  title: 'bonjour',
  fontWeight: '200',
  width: '325px',
  headingBorderBottom: '2px',
  headingBorderColor: 'black',
  messageInputHeight: '70px',
  buttonHoverColor: 'grey'
};

function Form(_ref) {
  var children = _ref.children,
      theme = _ref.theme;
  var methods = reactHookForm.useForm();

  var onSubmit = function onSubmit(data) {
    return console.log(data);
  };

  Object.assign(defaultTheme, theme);
  return React.createElement(styled.ThemeProvider, {
    theme: defaultTheme
  }, React.createElement(reactHookForm.FormProvider, Object.assign({}, methods), React.createElement(StyledForm, {
    className: 'App',
    onSubmit: methods.handleSubmit(onSubmit)
  }, children)));
}

function Heading(_ref) {
  var title = _ref.title;
  return React.createElement("div", {
    className: 'heading_wrapper'
  }, React.createElement(StyledHeading, null, " ", title || 'Get in touch!'));
}

function FirstName() {
  var _useFormContext = reactHookForm.useFormContext(),
      register = _useFormContext.register,
      errors = _useFormContext.errors;

  return React.createElement("div", {
    className: 'first_name_wrapper'
  }, React.createElement(StyledLabel, null, "First Name"), React.createElement(StyledInput, {
    name: 'firstName',
    ref: register({
      required: true
    })
  }), errors.firstName && React.createElement("p", null, "This field is required"));
}

function LastName() {
  var _useFormContext = reactHookForm.useFormContext(),
      register = _useFormContext.register,
      errors = _useFormContext.errors;

  return React.createElement("div", {
    className: 'last_name_wrapper'
  }, React.createElement(StyledLabel, null, "Last Name"), React.createElement(StyledInput, {
    name: 'lastName',
    ref: register({
      required: true
    })
  }), errors.lastName && React.createElement("p", null, "This field is required"));
}

function Email() {
  var _useFormContext = reactHookForm.useFormContext(),
      register = _useFormContext.register,
      errors = _useFormContext.errors;

  return React.createElement("div", {
    className: 'email_wrapper'
  }, React.createElement(StyledLabel, null, "Email Address"), React.createElement(StyledInput, {
    name: 'email',
    ref: register({
      required: true,
      pattern: {
        value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
        message: 'Invalid email address, please try again!'
      }
    })
  }), errors.email && errors.email.message, errors.email && errors.email.type === 'required' && React.createElement("p", null, "This field is required"));
}

function Message() {
  var _useFormContext = reactHookForm.useFormContext(),
      register = _useFormContext.register,
      errors = _useFormContext.errors;

  return React.createElement("div", {
    className: 'message_wrapper'
  }, React.createElement(StyledLabel, null, "Message"), React.createElement(StyledMessageInput, {
    name: 'message',
    className: 'message_input',
    ref: register({
      required: true,
      minLength: 55
    })
  }), errors.message && errors.message.type === 'required' && React.createElement("p", null, "This field is required"), errors.message && errors.message.type === 'minLength' && React.createElement("p", null, "More characters needed!"));
}

function SubmitButton(_ref) {
  var buttonTitle = _ref.buttonTitle;
  return React.createElement("div", {
    className: 'buttonTitle_wrapper'
  }, React.createElement(StyledButton, {
    type: 'submit'
  }, buttonTitle || 'Submit'));
}

exports.Email = Email;
exports.FirstName = FirstName;
exports.Heading = Heading;
exports.LastName = LastName;
exports.Message = Message;
exports.SubmitButton = SubmitButton;
exports.default = Form;
